[#50236] Allow sharing with groups

aaron-contreras opened the pull request:
## Description

### High-level demos of what's bundled in this PR

#### Given the following group setup
![Group setup](https://github.com/opf/openproject/assets/61627014/b0e050fa-502e-4bb4-a2de-bd75e33e0624)

#### 1. Including Groups in share modal's auto-completer search results
![Autocompleter demo](https://github.com/opf/openproject/assets/61627014/2b31a5b8-a8a6-4633-8fa9-d8bfcbb9f218)

#### 2. Only displaying the Group's share unless a specific Role is granted to a Group Member

![Displaying correct shares demo](https://github.com/opf/openproject/assets/61627014/ef602d14-d39e-44d9-8382-51309a37f86d)

#### 3. Updating a group's share also propagates to its inherited member roles
Hard to demonstrate on a screen recording but there's a feature spec :)

![Updating a group share demo](https://github.com/opf/openproject/assets/61627014/96c93890-bd63-46b8-8f15-33ce803a7bde)

#### 4. Removing a Group's share preserves the non-inherited shares a member might have
![Removing group shares demo](https://github.com/opf/openproject/assets/61627014/343c18ee-1d8d-47a2-9e53-8daed9cc2106)


#### 5. Mailer specific copy when sharing with Groups
![Mailer screenshot](https://github.com/opf/openproject/assets/61627014/aa2cea4d-1c6f-4092-892f-127e2ad89cb4)

## Notes
See: https://community.openproject.org/work_packages/50236

aaron-contreras committed:
Merge pull request #13855 from opf/bug/50392-public-permissions-selectable-on-role-administration

hide public permissions from role administration

aaron-contreras committed:
Include Groups in the auto-completer search results

aaron-contreras committed:
Promote `HavingEntityMembership` to the `Principal` level

Allows querying for entity memberships on groups as well as users.

aaron-contreras committed:
Lint: Address Rubocop warnings in `spec/models/principal_spec.rb`

* Add `frozen_string_literal: true` magic comment.

* Use `described_class` instead of the class constant.

* Use anonymous argument forwarding.

aaron-contreras committed:
Include Groups in shares list

Allow inviting a group and rendering turbo stream responses for Groups
as well as Users.

aaron-contreras committed:
Perform Inherited Role Creation on successful Create Service Call

aaron-contreras committed:
Lint: Address Rubocop offenses in `cleanup_inherited_roles_integration_spec.rb`

* Add `frozen_string_literal: true` magic comment.

* Use `contain_exactly` matcher instead of `match_array`

aaron-contreras committed:
Performance: Improve performance of integration spec

There is no case of factory cascading as demonstrated by the test-prof output
in the below ** BEFORE ** section. However, we're creating a lot of factories
and spending ~40% of the time in factories.

By using `shared_let`, we can reduce the number of factories being created
and speed up the test with the same behavior as it currently has.

Execution time was cut to < 1 second, meaning the spec file is now 2.6x faster.

** BEFORE **

Finished in 2.47 seconds (files took 9.5 seconds to load)
17 examples, 0 failures

Randomized with seed 8911

[TEST PROF INFO] Time spent in factories: 00:01.863 (40.63% of total time)
[TEST PROF INFO] Factories usage

 Total: 210
 Total top-level: 160
 Total time: 00:01.863 (out of 00:08.721)
 Total uniq factories: 9

   total   top-level     total time      time per call      top-level time               name

      50           0        0.0872s            0.0017s             0.0000s notification_setting
      34          34        0.4331s            0.0127s             0.4331s               user
      21          21        0.0600s            0.0029s             0.0600s               role
      21          21        0.0551s            0.0026s             0.0551s        global_role
      17          17        0.2831s            0.0167s             0.2831s              group
      17          17        0.0770s            0.0045s             0.0770s             member
      17          17        0.0518s            0.0030s             0.0518s      global_member
      17          17        0.7791s            0.0458s             0.7791s            project
      16          16        0.1244s            0.0078s             0.1244s              admin

** AFTER **

Finished in 0.95206 seconds (files took 8.81 seconds to load)
17 examples, 0 failures

Randomized with seed 27229

[TEST PROF INFO] Time spent in factories: 00:00.441 (14.48% of total time)
[TEST PROF INFO] Factories usage

 Total: 14
 Total top-level: 11
 Total time: 00:00.441 (out of 00:07.169)
 Total uniq factories: 9

   total   top-level     total time      time per call      top-level time               name

       3           0        0.0168s            0.0056s             0.0000s notification_setting
       2           2        0.0156s            0.0078s             0.0156s               user
       2           2        0.0121s            0.0061s             0.0121s        global_role
       2           2        0.0234s            0.0117s             0.0234s               role
       1           1        0.2310s            0.2310s             0.2310s            project
       1           1        0.0031s            0.0031s             0.0031s      global_member
       1           1        0.0440s            0.0440s             0.0440s              admin
       1           1        0.0872s            0.0872s             0.0872s              group
       1           1        0.0250s            0.0250s             0.0250s             member

aaron-contreras committed:
Extend cleanup inherited roles integration spec to cover work package roles

aaron-contreras committed:
Extend `WorkPackageMembers::DeleteService` with cleanup feature

aaron-contreras committed:
Allow sharing with a User despite the Group share it might already have

aaron-contreras committed:
Enqueue emails for Group shares

aaron-contreras committed:
Combine "Group" and "User" sharing specs together

aaron-contreras committed:
Ensure group users get their shares updated along with changes to a Group's share

aaron-contreras committed:
Lint: Address cop about usage of `match_array`

* Replaces `match_array` with `contain_exactly` where applicable.

aaron-contreras committed:
Performance: Improve performance of `update_roles_service_integration_spec`

We're spending over 50% of the time spent in factories. By leveraging the
`shared_let` strategy we can chop this down significantly to something more
reasonable.

Performance was improved by 50% as well as with a 50% reduction in time spent in
factories.

** Before **
Finished in 4.82 seconds (files took 9.09 seconds to load)
46 examples, 0 failures

Randomized with seed 3124

[TEST PROF INFO] Time spent in factories: 00:04.020 (57.97% of total time)
[TEST PROF INFO] Factories usage

 Total: 499
 Total top-level: 366
 Total time: 00:04.020 (out of 00:11.324)
 Total uniq factories: 9

   total   top-level     total time      time per call      top-level time               name

     133           0        0.2047s            0.0015s             0.0000s notification_setting
      88          88        0.9106s            0.0103s             0.9106s               user
      79          79        0.2084s            0.0026s             0.2084s       project_role
      50          50        0.6100s            0.0122s             0.6100s              group
      45          45        0.3730s            0.0083s             0.3730s              admin
      42          42        0.1811s            0.0043s             0.1811s             member
      38          38        1.6674s            0.0439s             1.6674s            project
      16          16        0.0382s            0.0024s             0.0382s        global_role
       8           8        0.0321s            0.0040s             0.0321s      global_member

** After **

Finished in 2.15 seconds (files took 8.89 seconds to load)
46 examples, 0 failures

Randomized with seed 44002

[TEST PROF INFO] Time spent in factories: 00:01.168 (27.67% of total time)
[TEST PROF INFO] Factories usage

 Total: 121
 Total top-level: 117
 Total time: 00:01.168 (out of 00:08.508)
 Total uniq factories: 9

   total   top-level     total time      time per call      top-level time               name

      50          50        0.5710s            0.0114s             0.5710s              group
      42          42        0.1891s            0.0045s             0.1891s             member
       9           9        0.0561s            0.0062s             0.0561s       project_role
       8           8        0.0338s            0.0042s             0.0338s      global_member
       4           0        0.0159s            0.0040s             0.0000s notification_setting
       3           3        0.0233s            0.0078s             0.0233s               user
       3           3        0.0101s            0.0034s             0.0101s        global_role
       1           1        0.0376s            0.0376s             0.0376s              admin
       1           1        0.2474s            0.2474s             0.2474s            project

aaron-contreras committed:
Add member role propagation expectations on feature spec

Comes with a slight performance improvement

ulferts commented on app/components/work_packages/share/modal_body_component.html.erb:
```diff
@@ -23,11 +23,13 @@
+            shared_principals.each do |principal|
+              share = principal.work_package_shares
+                               .where(entity: @work_package)
+                               .first
```
Not caused by this PR but we need to find a better way of dealing with this. The way we are currently doing it will lead to n+1 queries. 

Not part of the PR so does not need to be fixed in it.

aaron-contreras replied:
https://community.openproject.org/work_packages/50544/activity

ulferts commented on app/components/work_packages/share/permission_button_component.rb:
```diff
@@ -70,8 +70,14 @@ def options
+               .first
+        else
+          share.roles.first
+        end
```
Same as above, we need to avoid n+1 loading this information. But as above, it can also happen outside of this PR.

aaron-contreras replied:
https://community.openproject.org/work_packages/50544/activity

ulferts commented on app/components/work_packages/share/permission_button_component.rb:
```diff
@@ -70,8 +70,14 @@ def options
+          share.roles
+               .joins(:member_roles)
+               .where(member_roles: { inherited_from: nil })
+               .first
```
`share.roles` will already have `member_roles` joined as it is the join table between `members` and `roles`.

ulferts commented on app/controllers/work_packages/shares_controller.rb:
```diff
@@ -131,7 +131,11 @@ def find_role_ids(builtin_value)
+                                        .joins(:member_roles)
+                                        .of_work_package(@work_package)
+                                        .where(member_roles: { inherited_from: nil })
+                                        .size
```
In this PR you are redoing the pattern `Member.joins(:member_roles).where(member_roles: { inherited_from: nil))`. Additionally I see  `Role.join(:member_roles).where(member_roles: { inherited_from: nil))`.  

You could have a `MemberRole.not_inherited` scope. And by that have:
`Member.joins(:member_roles).where(MemberRole.not_inherited))`
`Role.join(:member_roles).where(MemberRole.not_inherited)`

You'd still end up repeating the joins. That can also be placed into a scope if you deem it appropriate. All of this is not super important admittedly.

ulferts commented on app/forms/work_packages/share/invitee.rb:
```diff
@@ -42,7 +42,7 @@ class Invitee < ApplicationForm
           resource: 'principals',
-          filters: [{ name: 'type', operator: '=', values: ['User'] },
+          filters: [{ name: 'type', operator: '=', values: %w[User Group] },
                     { name: 'id', operator: '!', values: [::Queries::Filters::MeValue::KEY] }],
```
There is an edge case hidden in here:

If the user is part of a group, that user can now give permissions to the group they are member of and by that in effect manage their own sharing. 

I don't think it is super critical for now since share_work_packages requires edit_work_packages. But this might be subject to change. And even now, the user potentially gains the ability to:
* be assignee 
* log own time and see them
* manage relations
over what the edit_work_package permission allows.

The only thing we could do to avoid this would be to limit the groups to those the user is not a member in and that might also be weird as it is hard to communicate that to the user. 

Or we exclude the user from being added when adding their group.

But ideas sound bad and I don't have better ones. One could also argue that we have the same problem with members where having the manage members permission allows to set additional permission for oneself.

ulferts replied:
I copied that question to the Share-Epic on OpenProject community.

ulferts commented on app/models/principal.rb:
```diff
@@ -57,12 +57,19 @@ class Principal < ApplicationRecord
+           inverse_of: :principal,
+           dependent: :delete_all,
+           class_name: 'Member',
+           foreign_key: 'user_id'
```
We now have:
```
has_many :members, dependent: :destroy
has_many :memberships, dependent: :nullify
has_many :work_package_shares, dependent: :delete_all
```

Which all reference the same entries in the database. I don't know which one wins.

An inherited problem but I would like to know what actually happens.

ulferts commented on app/workers/mails/work_package_shared_job.rb:
```diff
@@ -31,14 +31,44 @@ class Mails::WorkPackageSharedJob < ApplicationJob
+          .where(principal: work_package_member.principal.users)
+          .group('members.id')
+          .having("COUNT(member_roles.id) = 1")
+          .select('members.id')
```
Minor thing and not awfully relevant here but typically you want to use `COUNT(*)` over `COUNT(column)`. The only reason for using `COUNT(column)` is if you want to exclude NULL values which that statement filters out. But it has to check that value for doing it which is why it is slower.

ulferts commented on spec/features/work_packages/share_spec.rb:
```diff
@@ -82,66 +96,160 @@
-      # Reopening the modal will show the same state as before.
-      work_package_page.visit!
+      share_modal.close
+      click_button 'Share'
```
Why close the modal at this step?

aaron-contreras replied:
No reason. It's a left-over

ulferts commented on spec/features/work_packages/share_spec.rb:
```diff
@@ -82,66 +96,160 @@
+        # This user has never been added
+        share_modal.expect_not_shared_with(non_shared_project_user)
+
+        share_modal.expect_shared_count_of(7)
```
Kudos 🙇 for the love you showed on this feature spec.

aaron-contreras replied:
Thank you :)

ulferts commented on spec/support/components/work_packages/share_modal.rb:
```diff
@@ -78,7 +80,12 @@ def change_role(user, role_name)
+      # trick but this feels buggy.
       def close
+        modal_element.click
         modal_element.send_keys(:escape)
```
By now there is a close button on the share modal so we can just press that one. So good if the usability is right. 😄

ulferts reviewed (changes requested):
I like the PR. Reads straightforward and the added tests give a feeling of being solid.

When trying out the functionality a couple of scenarios fail. Since the code for fanning out and destroying group memberships wasn't touched, it might well be that the problem is caused by code that existed before and has had the same problem all the time.

### Scenario 1

Given: 
* User is part of the group

Steps:
* Add user with permission "View"
* Add group with permission "Edit"
* Edit user's share to have permission "Edit"
* Remove group's share

Expected
* Group's share is removed
* Users' share stays

Actual:
* Both shares are removed

### Scenario 2

Given: 
* User is part of the group

Steps:
* Add user with permission "View"
* Add group with permission "Edit"
* Edit user's share to have permission "Edit"
* Re-Add group to have permission "View"

Expected
* Group's share is updated
* Users' share stays the same

Actual:
* Group's share is updated
* User's share is removed

### Scenario 3

Given: 
* User is part of the group

Steps:
* Add group with permission "Edit"
* Add user with permission "Edit"

Expected
* Group's share is unchanged
* Users' share is added with "Edit" permission

Actual:
* Group's share is unchanged
* Users' share is is not added

### Scenario 4

Given: 
* User is part of the group

Steps:
* Add group with permission "Edit"
* Add user with permission "View"
* Update user's share to have permission "Edit"
* Close and reopen modal

Expected
* Group's share is showing "Edit"
* Users' share is showing "Edit"

Actual:
* Group's share is showing "Edit"
* Users' share vanished

aaron-contreras commented:
Hi @ulferts , thanks for catching those edge cases❗️ As discussed in our quick-chat this morning, the pattern indeed re-occurs when both the Group and User are granted the same Role and I have tracked down what's going on. I've detailed my findings with a walkthrough of what's happening and I've verified the current behavior of how Project memberships with Users and Groups are currently behaving and actually there is no bug coming from before, what was discovered in this PR is expected behavior (at least for Projects):

## Why is this happening?

The underlying issue lies in the `WorkPackageMembers::CreateService`, and `WorkPackageMembers::UpdateService`; specifically, in the `Members::Concerns::RoleAssignment` concern. Let's take the first mentioned scenario:

### Walkthrough

Given:
- User is part of the group

Steps:
 1. Add user with permission "View"
 2. Add group with permission "Edit"
 3. Edit user's share to have permission "Edit"
 4. Remove group's share
 
Expected:
- Group's share is removed
- User's share stays

Actual:
- Both shares are removed

**Step 1**

Sharing the work package with the user with the "View" role creates the following DB structure:

<img width="652" alt="image" src="https://github.com/opf/openproject/assets/61627014/db79ac56-eae2-449b-8392-b7f882c63128">

**Step 2**

Subsequently, sharing the work package with the group with the "Edit role" creates the following DB structure:

<img width="652" alt="image" src="https://github.com/opf/openproject/assets/61627014/ca00eaa9-2fb5-45e4-bc6e-0a16a6fc123d">


All is fine after this step. However, when executing...

**Step 3**

The PATCH request performed by the select dropdown is providing `Member1` (The `Member` record for Dinesh). What happens next in the `SetAttributesService` is the root cause of the problem. I've annotated the code snippets with a step by step walkthrough:
```ruby
module WorkPackageMembers
  class SetAttributesService < ::BaseServices::SetAttributes
    prepend Members::Concerns::RoleAssignment

    private

    def set_attributes(params)
      # The call to super invokes the `set_attributes` method in the prepended concern
      super

      model.change_by_system do
        model.project = model.entity&.project
      end
    end
  end
end
```

```ruby
module Members::Concerns::RoleAssignment
  private

  # Which invokes this method
  def set_attributes(params)
    assign_roles(params)

    super
  end

  def assign_roles(params)
    raise ArgumentError, 'Cannot handle changing `roles`. Use role_ids instead.' if model.persisted? && params.key?(:roles)

    return unless params[:role_ids]

    role_ids = params
                 .delete(:role_ids)
                 .select(&:present?)
                 .map(&:to_i)
    # role_ids is the id of "EditRole"
    
    existing_ids = model.member_roles.map(&:role_id)
    # Dinesh's roles via his member_roles are "EditRole" and "ViewRole"

    # ["EditRole", "ViewRole"] - ["EditRole"] => ["ViewRole"]
    # Which means Dinesh's MemberRole1 (the non-inherited one) gets
    # marked for destruction.
    mark_roles_for_destruction(existing_ids - role_ids)

    # ["EditRole"] - ["EditRole", "ViewRole"] => []
    # Which means that Dinesh is not getting a new "MemberRole4"
    # with the "EditRole" that is not inherited.
    # Here lies the bug!
    build_roles(role_ids - existing_ids)
  end

  def mark_roles_for_destruction(role_ids)
    role_ids.each do |role_id|
      model
        .member_roles
        .detect { |mr| mr.inherited_from.nil? && mr.role_id == role_id }
        &.mark_for_destruction
    end
  end

  def build_roles(role_ids)
    role_ids.each do |role_id|
      model.member_roles.build(role_id:)
    end
  end
end
```

And here it is. After performing **Step 3**, Dinesh ends up solely with the `MemberRole3` which he inherited from his Group

<img width="651" alt="image" src="https://github.com/opf/openproject/assets/61627014/dab1a681-6388-4e24-b860-ae804dbc49b1">


So then when performing **Step 4**, the cleanup job is doing its job properly and deleting `Member2` along with `MemberRole2` (Cool Group's `Member` record along with its `MemberRole` record), **AND** `Member1` along with its only left `MemberRole3` (which was inherited). That is why all shares for the work package are gone and this is correct behavior on the cleanup job's end.

The 3rd described scenario in the failing cases demonstrates exactly what's happening in the `SetAttributesService`:

**Scenario 3**

Given:
* User is part of the group

Steps:
* Add group with permission "Edit"
* Add user with permission "Edit"

Expected:
* Group's share is unchanged
* Users' share is added with "Edit" permission

Actual:
* Group's share is unchanged
* Users' share is is not added


The reason why this is behaving in this manner can be digested in the following two bullet points providing a digest of how Project memberships for Users and Groups have worked thus far:

Given "Role A" exists
Given you're a member of "Cool Group", "Cooler Group", and _n_ other number of groups.

* "Role A" can only ever be granted to you via _n_ groups you're a member of (each creating a separate `member_role` record in your Project membership with its inherited_from column`inherited_from: <group_member_role_id>`) **OR** as a direct membership in a project. Meaning if you were granted "Role A" independently and then any group you belong to is granted "Role A", you would only keep the inherited `member_role` with "Role A" just granted to you via the group. This sort of makes sense because why would you have the same role assigned from two different sources on the same resource. At least, I believe that was the train of thought back in time.
* Following up with the previous bullet point, you can be granted "Role A" multiple times **as long as it is granted via a different group**. Meaning if you are part of "Cool Group" and "Cooler Group" and both groups were granted "Role A", you would have two distinct `member_roles` each with a different `inherited_from` id (respectively, "Cool Group"'s `member_role` and "Cooler Group"'s `member_role`). That's how the `CleanupInheritedRolesService` is able to work its magic when you're removed from a group or when a group's membership is revoked, given that there are two separate `member_role`s granted to you by both groups.

Following all of the above, this raises the following questions:
1. Do we want to be able to grant the same Role to a User independently **AND** via any _n_ number of groups?
2. If so, which benefits does this yield?

I would find reasonable if this works exactly as described in the Project membership mechanics, but I would also find it somewhat reasonable to be able to grant the same Role both independently and via _n_ number of groups: on the one hand, it'd be nice that if I had specifically granted Dinesh "Role A", and I granted a group he belonged to "Role A", that upon modifying the Role that the group is shared the work package with that Dinesh's role is intact given I specifically granted him "Role A". However, it seems to me like the former is exactly what a sharer might **NOT** desire.

I see two options:
1. Allowing the described behavior in entity memberships.
2. Modifying the share modal behavior:
   * Hide/disable the roles in the dropdowns in both the row and invite form that have been granted to a User via a group.
   * Given "Role A", "Role B", "Role C", and "Dinesh" is a member in "Cool Group"
   * If "Cool Group" was invited with "Role A", when attempting to invite "Dinesh" individually, only "Role B" and "Role C" should be available roles. 
   * If "Dinesh" was invited with "Role A", and subsequently "Cool Group" is invited with "Role A", then "Dinesh"'s share row should be removed from the dropdown modal (this could be surprising, some indication of what happened should be made visible)

aaron-contreras commented:
Discussed this today with @klaustopher and we DO want to be able to grant the same role to a user independently and via a group. Converting back to a draft to account for this

aaron-contreras committed:
Extend `Members::Concerns::RoleAssignment` for Entity Membership use-case

We'd like to be able to assign the same role independently and inherited
via a group. For this, we're making the `Members::Concerns::RoleAssignment`'s
`#existing_ids` extensible and overriding it with entity-specific
behavior in `WorkPackageMembers::Concerns::RoleAssignment` and using this
new concern in the `WorkPackageMembers::SetAttributesService`.

aaron-contreras committed:
Remove unnecessary join with `member_roles` table

aaron-contreras committed:
Add `only_non_inherited` MemberRole scope

aaron-contreras committed:
Remove unnecessary close and re-open of share modal

aaron-contreras committed:
Use `COUNT(*)` instead of `COUNT(column_name)`

Thanks Jens!

aaron-contreras committed:
Use close button instead of hacking a way out of the modal

klaustopher reviewed (commented):
PR looks good from my side. I'll leave the final review for @ulferts as he has brought up some valid concern and I cannot 100% assess if they are fulfilled.

ulferts reviewed (approved):
Thanks for the detective work, @aaron-contreras. Feels good now and I did not run into any more weird cases.
